File: UPNP/gupnp.py
#!/usr/bin/env python3
import gi
gi.require_version('GUPnP', '1.0')
gi.require_version('GUPnPAV', '1.0')
from gi.repository import GLib, GObject, GUPnP, GUPnPAV

class UPnP_Finder:
    def __init__(self, loop, search):
        self.loop = loop
        self.search = search
        self.contents = []
        self.context = GUPnP.Context.new(None, None, 0)
        self.control_point = GUPnP.ControlPoint.new(self.context,"urn:schemas-upnp-org:device:MediaServer:1")
        self.control_point.connect("device-proxy-available", self.device_proxy_available_cb, loop)
        self.control_point.set_active(True)

    def action_browse(self, obj_id="0"):
        in_name = ["ObjectID","BrowseFlag","Filter","StartingIndex","RequestedCount", "SortCriteria"]
        in_val  = [obj_id, "BrowseDirectChildren", "*", "0", "0", ""]
        ou_name = ["Result"]
        ou_type = [GObject.TYPE_STRING.name]
        return self.content_directory.send_action_list(action="Browse", in_names=in_name, in_values=in_val, out_names=ou_name, out_types=ou_type)

    def print_server_info(self, proxy):
        print ("="*40)
        print (("Name     : %s")%(proxy.get_friendly_name()))
        print (("Location : %s")%(self.content_directory.get_location()))
        print (("udn      : %s")%(self.content_directory.get_udn()))
        print (("type     : %s")%(self.content_directory.get_service_type()))

    def device_proxy_available_cb(self, cp, proxy, data):
        self.content_directory = proxy.get_service("urn:schemas-upnp-org:service:ContentDirectory")

        ret, lst = self.action_browse()
        if ret :
            parse = GUPnPAV.DIDLLiteParser()
            parse.connect("container-available", self.container_available_cb, None)
            parse.parse_didl(lst[0])

        if (len(self.contents) != 0):
            self.loop.quit()

    def container_available_cb(self, parser, obj, data):
        if obj.get_child_count() > 0:
            ret, lst = self.action_browse(obj.get_id())
            if ret :
                parse = GUPnPAV.DIDLLiteParser()
                if obj.get_title().startswith(self.search):
                    parse.connect("container-available", self.container_found_cb, None)
                    parse.connect("item-available", self.item_found_cb, None)
                else:
                    parse.connect("container-available", self.container_available_cb, None)
                parse.parse_didl(lst[0])

    def container_found_cb(self, parser, obj, cd):
        if self.contents.count((obj.get_title(), ">")) == 0:
            self.contents.append((obj.get_title(), ">"))
        if obj.get_child_count() > 0:
            ret, lst = self.action_browse(obj.get_id())
            if ret :
                parse = GUPnPAV.DIDLLiteParser()
                parse.connect("container-available", self.container_found_cb, None)
                parse.connect("item-available", self.item_found_cb, None)
                parse.parse_didl(lst[0])
        if self.contents.count((obj.get_title(), "<")) == 0:
            self.contents.append((obj.get_title(), "<"))

    def item_found_cb(self, parser, obj, cd):
        if self.contents.count((obj.get_title(), obj.get_resources()[0].get_uri())) == 0:
            self.contents.append((obj.get_title(), obj.get_resources()[0].get_uri()))

    def __str__(self):
        ret = ""
        tab = 0
        for v1, v2 in self.contents:
            if v2 == '>':
                ret += ("%s%s:\n")%('\t'*tab,v1)
                tab = tab + 1
            elif v2 == '<':
                tab = tab - 1
            else:
                ret += ("%s%s\t%s\n")%('\t'*tab,v2,v1)
        return ret

def main():
    loop = GLib.MainLoop.new(None, False)
    fin = UPnP_Finder(loop, "SALMON")
    loop.run()
    print("="*40, "TEST CONTENTS", "="*40)
    print(fin)

if __name__ == "__main__":
    main()

File: UPNP/gupnp.c
#include <libgupnp/gupnp-control-point.h>

static GMainLoop *main_loop;

static void
device_proxy_available_cb (GUPnPControlPoint *cp,
                           GUPnPDeviceProxy *proxy,
                           gpointer           userdata)
{
  GUPnPServiceProxy *cd = GUPNP_SERVICE_PROXY(gupnp_device_info_get_service(GUPNP_DEVICE_INFO(proxy), "urn:schemas-upnp-org:service:ContentDirectory"));

  g_print("==============================\n");
  g_print("Name : %s\n", gupnp_device_info_get_friendly_name(GUPNP_DEVICE_INFO(proxy)));
  g_print("Location : %s\n", gupnp_service_info_get_location(GUPNP_SERVICE_INFO(cd)));
  g_print("udn : %s\n", gupnp_service_info_get_udn(GUPNP_SERVICE_INFO(cd)));
  g_print("type : %s\n", gupnp_service_info_get_service_type(GUPNP_SERVICE_INFO(cd)));



  const char *location;
  char *result = NULL;
  guint count, total;
  GError *error = NULL;

  location = gupnp_service_info_get_location (GUPNP_SERVICE_INFO (cd));

  g_print ("ContentDirectory available:\n");
  g_print ("\tlocation: %s\n", location);

  /* We want to be notified whenever SystemUpdateID (of type uint)
   * changes */
/*
  gupnp_service_proxy_add_notify (proxy,
          "SystemUpdateID",
          G_TYPE_UINT,
          notify_cb,
          (gpointer) "Test");
*/

  /* Subscribe */
/*
  g_signal_connect (proxy,
          "subscription-lost",
          G_CALLBACK (subscription_lost_cb),
          NULL);

  gupnp_service_proxy_set_subscribed (proxy, TRUE);
*/

  /* And test action IO */
  gupnp_service_proxy_send_action (cd,
          "Browse",
          &error,
          /* IN args */
          "ObjectID",
          G_TYPE_STRING,
          "0",
          "BrowseFlag",
          G_TYPE_STRING,
          "BrowseDirectChildren",
          "Filter",
          G_TYPE_STRING,
          "*",
          "StartingIndex",
          G_TYPE_UINT,
          0,
          "RequestedCount",
          G_TYPE_UINT,
          0,
          "SortCriteria",
          G_TYPE_STRING,
          "",
          NULL,
          /* OUT args */
          "Result",
          G_TYPE_STRING,
          &result,
          "NumberReturned",
          G_TYPE_UINT,
          &count,
          "TotalMatches",
          G_TYPE_UINT,
          &total,
          NULL);

  if (error) {
      g_printerr ("Error: %s\n", error->message);
      g_error_free (error);

      return;
  }

  g_print ("Browse returned:\n");
  g_print ("\tResult:         %s\n", result);
  g_print ("\tNumberReturned: %u\n", count);
  g_print ("\tTotalMatches:   %u\n", total);

  g_free (result);
  //g_main_loop_quit(main_loop);
}

int
main (int argc, char **argv)
{
  GUPnPContext *context;
  GUPnPControlPoint *cp;

  /* Required initialisation */
#if !GLIB_CHECK_VERSION(2,35,0)
  g_type_init ();
#endif

  /* Create a new GUPnP Context.  By here we are using the default GLib main
   *      context, and connecting to the current machine's default IP on an
   *           automatically generated port. */
  context = gupnp_context_new (NULL, NULL, 0, NULL);

  /* Create a Control Point targeting WAN IP Connection services */
  cp = gupnp_control_point_new
    (context, "urn:schemas-upnp-org:device:MediaServer:1");

  /* The service-proxy-available signal is emitted when any services which match
   *      our target are found, so connect to it */
  g_signal_connect (cp,
      "device-proxy-available",
      G_CALLBACK (device_proxy_available_cb),
      NULL);

  /* Tell the Control Point to start searching */
  gssdp_resource_browser_set_active (GSSDP_RESOURCE_BROWSER (cp), TRUE);

  /* Enter the main loop. This will start the search and result in callbacks to
   *      service_proxy_available_cb. */
  main_loop = g_main_loop_new (NULL, FALSE);
  g_main_loop_run (main_loop);

  /* Clean up */
  g_main_loop_unref (main_loop);
  g_object_unref (cp);
  g_object_unref (context);

  return 0;
}
